Sources/NIOWebSocket/NIOWebSocketFrameAggregator.swift

```diff
@@ -0,0 +1,136 @@
+    ) {
+        self.minNonFinalFragmentSize = minNonFinalFragmentSize
+        self.maxAccumulatedFrameCount = maxAccumulatedFrameCount
+        self.maxAccumulatedFrameSize = maxAccumulatedFrameSize
+    }
+    
+    
+    public func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        let frame = unwrapInboundIn(data)
+        do {
+            switch frame.opcode {
+            case .continuation:
+                guard let firstFrameOpcode = self.bufferedFrames.first?.opcode else {
+                    throw Error.didReceiveFragmentBeforeReceivingTextOrBinaryFrame
+                }
+                try self.bufferFrame(frame)
+                
+                guard frame.fin else { break }
+                // final frame received
+                
+                let aggregatedFrame = self.aggregateFrames(opcode: firstFrameOpcode)
+                self.clearBuffer()
+                
+                context.fireChannelRead(wrapInboundOut(aggregatedFrame))
+            case .binary, .text:
+                if frame.fin {
+                    guard self.bufferedFrames.isEmpty else {
+                        throw Error.receivedNewFrameWithoutFinishingPrevious
+                    }
+                    // fast path: no need to check any constraints nor unmask and copy data
+                    context.fireChannelRead(data)
+                } else {
+                    try self.bufferFrame(frame)
+                }
+            default:
+                // control frames can't be fragmented
+                context.fireChannelRead(data)
+            }
+        } catch {
+            // free memory early
+            self.clearBuffer()
+            context.fireErrorCaught(error)
+        }
+    }
+    
+    private func bufferFrame(_ frame: WebSocketFrame) throws {
+        guard self.bufferedFrames.isEmpty || frame.opcode == .continuation else {
+            throw Error.receivedNewFrameWithoutFinishingPrevious
+        }
+        guard frame.fin || frame.length >= self.minNonFinalFragmentSize else {
+            throw Error.nonFinalFragmentSizeIsTooSmall
+        }
+        guard self.bufferedFrames.count < self.maxAccumulatedFrameCount else {
+            throw Error.tooManyFragments
+        }
+        
+        // if this is not a final frame, we will at least receive one more frame
+        guard frame.fin || (self.bufferedFrames.count + 1) < self.maxAccumulatedFrameCount else {
+            throw Error.tooManyFragments
+        }
+        
+        self.bufferedFrames.append(frame)
+        self.accumulatedFrameSize += frame.length
+        
+        guard self.accumulatedFrameSize <= self.maxAccumulatedFrameSize else {
+            throw Error.accumulatedFrameSizeIsTooLarge
+        }
+    }
+    
+    private func aggregateFrames(opcode: WebSocketOpcode) -> WebSocketFrame {
+        var dataBuffer = ByteBuffer()
+        dataBuffer.reserveCapacity(minimumWritableBytes: self.accumulatedFrameSize)
```

Comment:
Since we are in the Channel pipeline, do we maybe want to use the channel's byteBufferAllocator for this?

Reply:
We do indeed, good catch @fabianfett.

Reply:
In principle a minor optimisation available to us is to use the buffer from the first frame to accumulate the rest, saving a copy. That also avoids the need to deal with this issue.